python/ray/air/execution/_internal/event_manager.py

```diff
@@ -163,40 +234,452 @@ def wait(
+            )
+            all_task_futures.extend(tracked_actor_task_futures)
+
+        # Stop / resource futures should have priority, but task futures should
+        # be shuffled.
+        random.shuffle(all_task_futures)
+        all_futures = stop_futures + resource_futures + all_task_futures
+
+        if num_events:
+            # If we have a fixed number of events, subtract the number of
+            # started/stopped actors as they count as an event.
+            num_returns = num_events - started_actors - stopped_actors
+        else:
+            # Otherwise, await all futures.
+            num_returns = len(all_futures)
+
+        assert num_returns >= 0
+
+        # If we have no events left (e.g. no futures or too many actor events), exit
+        if num_returns == 0:
+            return
+
+        # We can't wait for more events than we have futures
+        if num_returns > len(all_futures):
+            if num_returns <= len(self._tracked_actors_to_attrs):
+                # In this case, there are no resource futures, but we actually waited
+                # for some actors to start (without resource futures). So this is
+                # not an error - there's just nothing to actually await. So we return.
+                return
+
+            # Otherwise, we specified to wait for more events than we track, which is
+            # a usage error.
+            raise ValueError(
+                f"num_events cannot be greater than the number of events "
+                f"currently pending by the RayEventManager. "
+                f"Got num_events={num_returns} and {len(all_futures)} tracked events "
+                f"for event_type={event_type}."
+            )
+
+        # Here we await the futures
+        ready, not_ready = ray.wait(
+            all_futures, num_returns=num_returns, timeout=timeout
+        )
+
+        num_processed_events = len(ready) + stopped_actors + started_actors
+
+        # Process resolved futures
+        for future in ready:
+            if future in self._futures_to_tracked_actor_tasks:
+                # Actor task future
+                self._handle_ready_actor_task_future(future)
+            elif future in self._futures_to_actor_termination:
+                # Termination future resolved
+                tracked_actor = self._futures_to_actor_termination[future]
+                self._try_stop_tracked_actor(tracked_actor)
+            else:
+                # Resource future
+                self._handle_ready_resource_future()
+
+        # Lastly, if we have events left, try to start actors again (some resources
+        # might have been freed by our previous stop events).
+        if event_type in [EventType.ALL, EventType.ACTORS]:
+            if num_processed_events > 0 and num_events:
+                events_left = num_events - num_processed_events
+            else:
+                events_left = num_events
+
+            self._try_start_actors(max_actors=events_left)
```

Comment:
why tie events <-> actors ... this comment of "when we have events left, let's start actors" is really confusing :)